Lab3.sc

```diff
@@ -0,0 +1,154 @@
+import scala.annotation.tailrec
+
+//TASK 1
+
+def divideListWithNegativeElements(list:List[Int]):(List[Int],List[Int]) = {
+  def getNegativeList(x:List[Int]):List[Int]=
+    if(x==List())
+      List()
+    else if(x.head<0)
+      x.head :: getNegativeList(x.tail)
+    else
+      getNegativeList(x.tail)
+
+  def getOddList(y:List[Int]):List[Int] =
+    if(y==List())
+      List()
+    else if(y.head<0 && y.head%2!=0)
+      y.head :: getOddList(y.tail)
+    else
+      getOddList(y.tail)
+  (getNegativeList(list),getOddList(list))
+}
+
+divideListWithNegativeElements(List(-3,-6,8,-9,13))
+divideListWithNegativeElements(List(2,3,8,1,13))
+divideListWithNegativeElements(List(-3,-2,-11))
+divideListWithNegativeElements(List())
+divideListWithNegativeElements(List(2))
+divideListWithNegativeElements(List(-5))
+divideListWithNegativeElements(List(-6))
+
+//TASK 2
+
+def listLength[A](list:List[A]):Int =
+  {
+    @tailrec
+    def iter[B](x:List[B],len:Int): Int =
+      if(x==List())
+        len
+      else
+        iter(x.tail,len+1)
+
+    iter(list,0)
+  }
+listLength(List())
+listLength(List(1,2,3,4))
+listLength(List(List('a','b'),1,4))
+
+//zlozonosc obliczeniowa - dlugosc listy / zlozonosc pamieciowa - ramka funkci listLength
+
+//TASK 3
+
+def concatenateListsAlternately[A](left:List[A],right:List[A]):List[A] =
+  (left,right) match
+  {
+    case (Nil,Nil) => Nil
+    case  (Nil,r) => r
+    case  (l,Nil) =>  l
+    case  (lhead::ltail,rhead::rtail) => lhead::rhead::concatenateListsAlternately(left.tail,right.tail)
+  }
+
+concatenateListsAlternately(List(5,4,3,2),List(1,2,3,4,5,6))
+concatenateListsAlternately(List(1,2,3,4,5,6),List(5,4,3,2))
+concatenateListsAlternately(Nil,List(1,2,3))
+concatenateListsAlternately(List(1,2,3),Nil)
+concatenateListsAlternately(List(1,3,5),List(2,4,6))
+
+//zlozonosc pamieciowa - wszystkie wywolania funkcji / zlozonosc obliczeniowa - suma dlugosci obu list
```

Comment:
"- wszystkie wywolania funkcji" co to za opisowa analiza złożoności? To tak jakbym napisał Ci: ocena - taka jaką dostałeś ;)

Reply:
i w złożoności obliczeniowej masz błąd- jest to długość krótszej z list. Po jej przejściu reszta się dokleja. Swoją drogą doklejenie listy operatorem ::  (?)